### PR TITLE
New strategy : Retain up from polygon

### PR DESCRIPTION
**Pull request by tisseo-deploy**
**Summary:**

Solve Issue https://github.com/OneBusAway/onebusaway-gtfs-modules/issues/237 :  Add RetainUpFromPolygon strategy

### CODE REVIEW

File: docs/onebusaway-gtfs-transformer-cli.md [hunk 1]
```diff
@@ -270,6 +271,26 @@ You can remove a specific entity from a feed.
 
 Note that removing an entity has a cascading effect.  If you remove a trip, all the stop times that depend on that
 trip will also be removed.  If you remove a route, all the trips and stop times for that route will be removed.
+
+#### Retain Up From Polygon
```
Comment (leonardehrenfried) [inline docs/onebusaway-gtfs-transformer-cli.md:275]: I must admit that I find the terminology of "retaining up" and "retaining down" confusing. What do you mean by that?

File: onebusaway-gtfs-transformer/src/main/java/org/onebusaway/gtfs_transformer/impl/RetainUpFromPolygon.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return this.getClass().getSimpleName();
+    }
+
+    @Override
+    public void run(TransformContext transformContext, GtfsMutableRelationalDao gtfsMutableRelationalDao) {
+        Geometry geometry = buildPolygon(polygon);
```
Comment (leonardehrenfried) [inline onebusaway-gtfs-transformer/src/main/java/org/onebusaway/gtfs_transformer/impl/RetainUpFromPolygon.java:41]: Move this into the setter and check if it's valid and non-empty there. Throw an exception immediately if it isn't.

File: onebusaway-gtfs-transformer/src/main/java/org/onebusaway/gtfs_transformer/impl/RetainUpFromPolygon.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        * return null if an exception is encountered when parsing the wkt string
+     */
+    private Geometry buildPolygon(String polygonWKT) {
+        WKTReader reader = new WKTReader();
+        try{
+            return  reader.read(polygonWKT);
+        } catch (ParseException e){
+            String message = String.format("Error parsing WKT string : %s", e.getMessage());
+            log.error(message);
+            return null;
```
Comment (leonardehrenfried) [inline onebusaway-gtfs-transformer/src/main/java/org/onebusaway/gtfs_transformer/impl/RetainUpFromPolygon.java:79]: Don't cover up a bad input string and re-throw as a RuntimeException. The program should fail as soon as possible.

File: docs/onebusaway-gtfs-transformer-cli.md (hunk 1 above)
Comment (enrikomic) [reply docs/onebusaway-gtfs-transformer-cli.md:275]: We did'nt found a better term. This one is direct from the existing retain algotithm that seems very stable.
"Up" means that the algorithm will traverse up the entity dependency tree by applying retainUp to each of them. Starting from the stop, retainUp will be applied to the stop_times referencing this stop, then to the trips, and so on.
Then, when the base of the tree is reached, it applies retainDown to all the traversed entities. Therefore, all the trips of the route and then all the stop_times of each trip will be tagged as "retain".
Entities not marked as "retain" will be deleted.
The goal of the strategy is to retain all entities that are directly or indirectly linked to the area.

We will add this brief explanation to the documentation.

Comment (leonardehrenfried) [reply docs/onebusaway-gtfs-transformer-cli.md:275]: okay, that's fine.

**Comment by leonardehrenfried** [comment]
I will merge this and make a few small changes.

**Pull request merged**